server/src/main/java/ru/spbau/blackout/server/RoomServer.java

```diff
@@ -3,51 +3,67 @@
 import java.net.ServerSocket;
 import java.net.Socket;
+import java.util.ArrayList;
 import java.util.Deque;
+import java.util.List;
 import java.util.concurrent.ConcurrentLinkedDeque;
 import java.util.concurrent.atomic.AtomicInteger;
 
-public class RoomServer {
+class RoomServer {
+
+    private static final int PLAYERS_NUMBER = 2;
 
     private final int port;
-    private final Deque<RoomClientThread> roomClientThreads = new ConcurrentLinkedDeque<>();
+    private final Deque<ClientThread> clientThreads = new ConcurrentLinkedDeque<>();
     private final AtomicInteger playersNumber = new AtomicInteger();
 
-    public AtomicInteger getPlayersNumber() {
-        return playersNumber;
+     RoomServer(int port) {
+        this.port = port;
     }
 
-    public RoomServer(int port) {
-        this.port = port;
+    int getPlayersNumber() {
+        return playersNumber.get();
     }
 
-    public void run() {
+    @SuppressWarnings("InfiniteLoopStatement")
+    void run() {
         try (ServerSocket serverSocket = new ServerSocket(port)) {
-            while (true) {
+            do {
                 final Socket nextSocket = serverSocket.accept();
-                final RoomClientThread nextThread = new RoomClientThread(this, nextSocket);
-                synchronized (System.out) {
-                    System.out.println("New thread for a connection is created");
-                }
-                roomClientThreads.add(nextThread);
+                final ClientThread nextThread = new ClientThread(this, nextSocket);
+                log("New thread for a connection is created.");
+                clientThreads.add(nextThread);
                 playersNumber.addAndGet(1);
                 nextThread.start();
-            }
+                maybePlayGame(PLAYERS_NUMBER);
+            } while (true);
         } catch (IOException e) {
-            synchronized (System.out) {
-                System.out.println("Exception caught when trying to listen on port "
-                        + port + " or listening for a connection");
-                System.out.println(e.getMessage());
+            log("Exception caught when trying to listen on port " + port +
+                " or listening for a connection:" + e.getMessage());
+        }
+    }
+
+    private synchronized void maybePlayGame(int playersForGame) {
+        if (playersNumber.get() >= playersForGame) {
+            playersNumber.addAndGet(-playersForGame);
+            final List<ClientThread> clients = new ArrayList<>(playersForGame);
+            for (int playerIndex = 0; playerIndex < playersForGame; playerIndex++) {
+                clients.add(clientThreads.removeFirst());
             }
+            final Thread game = new Game(this, clients);
+            game.start();
         }
     }
 
-    void discard(RoomClientThread clientThread) {
-        roomClientThreads.remove(clientThread);
-        playersNumber.addAndGet(-1);
+    void discard(ClientThread clientThread) {
```

Comment:
The right way would be to not let concurrency slip through the borders of this class. Client threads shouldn't know that they are being managed by a concurrent server (or by any server). Ideally, they would not call discard(this) at all. It's a server's job to watch for threads and, if a thread exits, to decrement the counter.

How you could achieve that:
1) make ClientThread a Runnable
2) when there is a new client connecting:

new Thread( new Runnable() {
    @Override
    public void run() {
        try { clientRunnable.run(); } 
        finally { <do cleanup, or, better, notify the main thread that cleanup is necessary > }   
   }
}).start();

Doing the cleanup in the main thread would let you avoid any Concurrent.* and Atomic.* for which it is very hard to prove that they are being used properly.